test/python/synthesis/test_multi_controlled_synthesis.py

```diff
@@ -0,0 +1,384 @@
+        unitary gates.
+        """
+        base_gate = UGate(0.123, 0.456, 0.789)
+        qc = QuantumCircuit(num_ctrl_qubits + 1)
+        qc.append(base_gate.control(num_ctrl_qubits), range(num_ctrl_qubits + 1))
+        pm = generate_preset_pass_manager(
+            optimization_level=0, basis_gates=["u", "cx"], seed_transpiler=12345
+        )
+        transpiled_circuit = pm.run(qc)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+        # The synthesis of MCX(n) uses two MCRZ(n), one MCRY(n), and one MCPhase(n-1).
+        # Thus the number of CX-gate should be upper-bounded by
+        # 3*(16 * (n + 1) - 40) + (8 * (n-1)^2 - 16 * (n-1))
+        self.assertLessEqual(cx_count, 8 * num_ctrl_qubits**2 + 16 * num_ctrl_qubits - 96)
+
+    @combine(
+        num_ctrl_qubits=[1, 2, 3, 4, 5, 6, 7, 8],
+        base_gate=[
+            XGate(),
+            YGate(),
+            ZGate(),
+            HGate(),
+            PhaseGate(0.123),
+            SGate(),
+            SdgGate(),
+            TGate(),
+            TdgGate(),
+            SXGate(),
+            SXdgGate(),
+            RXGate(0.789),
+            RYGate(0.123),
+            RZGate(0.456),
+            UGate(0.1, 0.2, 0.3),
+        ],
+    )
+    def test_small_mc_gates_cx_count(self, num_ctrl_qubits, base_gate):
+        """Test that transpiling various multi-controlled gates with small number of controls and no
+        ancillas yields the expected number of CX gates.
+
+        This test prevents making changes to the synthesis algorithms that would deteriorate the
+        quality of the synthesized circuits.
+        """
+        qc = QuantumCircuit(num_ctrl_qubits + 1)
+        qc.append(base_gate.control(num_ctrl_qubits), range(num_ctrl_qubits + 1))
+        pm = generate_preset_pass_manager(
+            optimization_level=0, basis_gates=["u", "cx"], seed_transpiler=12345
+        )
+        transpiled_circuit = pm.run(qc)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+
+        if isinstance(base_gate, (XGate, YGate, ZGate, HGate)):
+            # MCX gate and other locally equivalent multi-controlled gates
+            expected = {1: 1, 2: 6, 3: 14, 4: 36, 5: 84, 6: 140, 7: 220, 8: 324}
+        elif isinstance(base_gate, (PhaseGate, SGate, SdgGate, TGate, TdgGate, SXGate, SXdgGate)):
+            # MCPhase gate and other locally equivalent multi-controlled gates
+            expected = {1: 2, 2: 6, 3: 20, 4: 44, 5: 84, 6: 140, 7: 220, 8: 324}
+        elif isinstance(base_gate, RZGate):
+            expected = {1: 2, 2: 4, 3: 14, 4: 24, 5: 40, 6: 56, 7: 80, 8: 104}
+        elif isinstance(base_gate, (RXGate, RYGate)):
+            expected = {1: 2, 2: 8, 3: 20, 4: 24, 5: 40, 6: 56, 7: 80, 8: 104}
+        elif isinstance(base_gate, UGate):
+            expected = {1: 2, 2: 22, 3: 54, 4: 92, 5: 164, 6: 252, 7: 380, 8: 532}
+        else:
+            # make sure we don't miss any case (and make pytint happy)
```

Comment:
Could we just raise an error instead here? Otherwise there could be a scenario where this silently passes (where some magic gate has 0 CX) -- in case you want to keep it as is, there's a small typo "pylint" couldn't catch 😛 
```suggestion
            # make sure we don't miss any case (and make pylint happy)
```

Reply:
Lol about me misspelling pylint and pylint not catching that. Raising an error is indeed better.